src/Microsoft.OData.Client/DeepInsertSaveResult.cs

```diff
@@ -225,19 +268,259 @@ protected override ODataRequestMessageWrapper CreateRequestMessage(string method
+            {
+                return new MaterializeAtom(responseInfo, new[] { materializerEntry.Entry }, entityDescriptor.Entity.GetType(), materializerEntry.Format, base.MaterializerCache, false);
+            }
+
+            return null;
         }
 
         protected override void HandleOperationResponse(IODataResponseMessage responseMessage)
         {
-            throw new NotImplementedException();
+            this.batchResponseMessage = responseMessage;
         }
 
         protected override DataServiceResponse HandleResponse()
         {
-            throw new NotImplementedException();
+            // This will process the responses and throw if failure was detected
+            if (this.ResponseStream != null)
+            {
+                return this.HandleDeepInsertResponse();
+            }
+
+            return new DataServiceResponse(null, (int)WebExceptionStatus.Success, new List<OperationResponse>(0), batchResponse: true);
+        }
+
+        private DataServiceResponse HandleDeepInsertResponse()
+        {
+            EntityDescriptor entityDescriptor = this.bulkUpdateGraph.TopLevelDescriptors[0];
+            MaterializerEntry entry = default;
+            DataServiceResponse response = new DataServiceResponse(headers, (int)this.batchResponseMessage.StatusCode, operationResponses, this.IsBatchRequest);
+
+            try
+            {
+                // We are setting the MaxProtocolVersion to v4.01 when reading deep insert responses.
+                this.RequestInfo.Context.MaxProtocolVersion = ODataProtocolVersion.V401;
+                Version responseVersion;
+                HandleResponse(
+                    this.RequestInfo,
+                    (HttpStatusCode)this.batchResponseMessage.StatusCode,
+                    this.batchResponseMessage.GetHeader(XmlConstants.HttpODataVersion),
+                    () => this.ResponseStream,
+                    throwOnFailure: true,
+                    out responseVersion);
+
+                HeaderCollection headers = new HeaderCollection(this.batchResponseMessage);
+
+                ResponseInfo responseInfo = this.CreateResponseInfo(entityDescriptor);
+                HttpWebResponseMessage responseMessageWrapper = new HttpWebResponseMessage(
+                    headers,
+                    this.batchResponseMessage.StatusCode,
+                    () => this.ResponseStream);
+
+                ODataMaterializerContext materializerContext = new ODataMaterializerContext(responseInfo, base.MaterializerCache);
+                entry = ODataReaderEntityMaterializer.ParseSingleEntityPayload(responseMessageWrapper, responseInfo, entityDescriptor.Entity.GetType(), materializerContext);
+
+                this.headers = headers;
+                this.topLevelEntry = entry;
+            }
+            catch (DataServiceRequestException)
+            {
+                throw;
+            }

```

Comment:
Do we need this block? We just catch the `DataServiceRequestException` and then rethrow therefore I don't see what purpose it serves